### PR TITLE
 #8540: Upgrade eltwise binary ops to support queue_id /output_tensor / uint output dtype

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
[Eltwise binary EQ should produce a boolean or integer tensor as output #8540](https://github.com/tenstorrent/tt-metal/pull/9029)
[Add Q_ID to ElemwiseBinary/ELEMWISE_BINARY_EQ #9109
](https://github.com/tenstorrent/tt-metal/issues/9109)

### What changed
After this change ttnn's eltwise binary ops like `eq`
- respect `output_dtype` or `output_tensor`'s dtype. In case uint32 return type requested, kernel calls sfpu typecast.
- supports ouput_tensor
- supports queue_id

#### Example 1: uint32 dtype
Eq returns bfloat by default, but if you pass a dtype the result will be correctly formed with 0 and 1.
```
result = ttnn.eq(input_tensor_a, input_tensor_b, dtype=ttnn.DataType.UINT32)
```
Same works if you call `ttnn::eq` or any other eltwise binary from C++

#### Example 2: output_tensor 
```
output_tensor_preallocated = ttnn.ones([h, w], output_dtype, ttnn.TILE_LAYOUT, device, ttnn.L1_MEMORY_CONFIG)
ttnn.eq(input_tensor_a, input_tensor_b, dtype=output_dtype, output_tensor=output_tensor_preallocated)    
```

#### queue_id example
```
result = ttnn.eq(1, input_tensor_a, input_tensor_b)
```

### Why
There is a request to support integer or boolean return type in eq. 

### Caution
This does not work on Greyskull
Most operations do not support integer inputs for now.
Typecast from int to bfloat is not implemented yet either.
It means that the use of the int tensor is limited.

### CODE REVIEW

File: build_metal.sh [hunk 1]
```diff
@@ -94,7 +94,7 @@ else
 fi
 
 echo "Building tt-metal"
-cmake_args="-B build -G Ninja -DCMAKE_CXX_COMPILER=clang++-17 -DCMAKE_EXPORT_COMPILE_COMMANDS=$export_compile_commands"
+cmake_args="-B build -G Ninja -DCMAKE_CXX_COMPILER=g++ -DCMAKE_EXPORT_COMPILE_COMMANDS=$export_compile_commands"
```
Comment (ayerofieiev-tt) [inline build_metal.sh:97]: Temporary

Comment (ayerofieiev-tt) [reply build_metal.sh:97]: Can't get clang based build to work with lldb or dbg debugger 😢

Comment (tt-rkim) [reply build_metal.sh:97]: Sounds good, pls delete once you're done

Or you can also add the option for compiler for us :)

File: tt_eager/tt_dnn/op_library/eltwise_binary/eltwise_binary_op.hpp [hunk 2]
```diff
@@ -132,14 +134,20 @@ struct make_eltwise_binary {
                     (in_a.get_legacy_shape() == in_b.get_legacy_shape()) or
                     (in_a.get_legacy_shape().without_padding() == in_b.get_legacy_shape().without_padding()),
                     "Input shapes must be the same!");
-                return operation::run(
+
+                DataType dtype = output_dtype.value_or(in_a.get_dtype());
```
Comment (ayerofieiev-tt) [inline tt_eager/tt_dnn/op_library/eltwise_binary/eltwise_binary_op.hpp:138]: Will find a way to retrieve dtype in a more abstract way to not mention specific op type here.

Comment (ayerofieiev-tt) [inline tt_eager/tt_dnn/op_library/eltwise_binary/eltwise_binary_op.hpp:138]: Need to remove this whole file

File: ttnn/cpp/ttnn/op_library/binary/binary_op.hpp [hunk 3]
```diff
@@ -124,15 +124,20 @@ struct ExecuteBinary {
             input_tensor_b = tt::tt_metal::repeat(input_tensor_b, repeats.value(), output_memory_config);
         }
 
-        return operation::run(
+        DataType dtype = output_dtype.value_or(input_tensor_a.get_dtype());
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/op_library/binary/binary_op.hpp:127]: Need to find a way to retrieve dtype in a more abstract way to not mention specific op type here before merging.

**Review by xanderchin — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Review by SeanNijjar — APPROVED** [review]

File: ttnn/cpp/pybind11/operations/binary.hpp [hunk 4]
```diff
@@ -51,34 +53,47 @@ void bind_binary_operation(py::module& module, const binary_operation_t& operati
...
                const ttnn::Tensor& input_tensor_a,
                const float scalar,
                const std::optional<ttnn::MemoryConfig>& memory_config,
                const std::optional<const DataType>& dtype,
-               const std::optional<std::vector<std::string>>& activations) -> ttnn::Tensor {
-                return self(input_tensor_a, scalar, memory_config, dtype, activations);
+               const std::optional<ttnn::Tensor>& output_tensor,
+               const std::optional<std::vector<std::string>>& activations,
+               const uint8_t& queue_id) -> ttnn::Tensor {
+                return self(queue_id, input_tensor_a, scalar, memory_config, dtype, output_tensor, activations);
```
Comment (eyonland) [inline ttnn/cpp/pybind11/operations/binary.hpp:66]: I thought we were going to make the queue_id in the front here?  Why is it the last arg?

Comment (ayerofieiev-tt) [reply ttnn/cpp/pybind11/operations/binary.hpp:66]: There are 2 versions in C++ for each of eq(tensor_a, tensor_b) vs eq(tensor_a, scalar): 
- with queue_id in front
- w/o

There is a single version in Py for each 
- via kwarg

**Comment by ayerofieiev-tt** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/9377463343 😃

**Pull request merged**